In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/ieee-fraud-detection/sample_submission.csv
/kaggle/input/ieee-fraud-detection/test_identity.csv
/kaggle/input/ieee-fraud-detection/train_identity.csv
/kaggle/input/ieee-fraud-detection/test_transaction.csv
/kaggle/input/ieee-fraud-detection/train_transaction.csv


**Data Preparation**

In [2]:
# Read data csv
train_transaction = pd.read_csv('../input/ieee-fraud-detection/train_transaction.csv', index_col='TransactionID')
test_transaction = pd.read_csv('../input/ieee-fraud-detection/test_transaction.csv', index_col='TransactionID')

train_identity = pd.read_csv('../input/ieee-fraud-detection/train_identity.csv', index_col='TransactionID')
test_identity = pd.read_csv('../input/ieee-fraud-detection/test_identity.csv', index_col='TransactionID')

#sample_submission = pd.read_csv('../input/ieee-fraud-detection/sample_submission.csv', index_col='TransactionID')


In [3]:
# Menampilkan lima baris pertama data train_transaction
train_transaction.head()

isFraud  TransactionDT  TransactionAmt ProductCD  card1  card2  \
TransactionID                                                                   
2987000              0          86400            68.5         W  13926    NaN   
2987001              0          86401            29.0         W   2755  404.0   
2987002              0          86469            59.0         W   4663  490.0   
2987003              0          86499            50.0         W  18132  567.0   
2987004              0          86506            50.0         H   4497  514.0   

               card3       card4  card5   card6  ...  V330  V331  V332  V333  \
TransactionID                                    ...                           
2987000        150.0    discover  142.0  credit  ...   NaN   NaN   NaN   NaN   
2987001        150.0  mastercard  102.0  credit  ...   NaN   NaN   NaN   NaN   
2987002        150.0        visa  166.0   debit  ...   NaN   NaN   NaN   NaN   
2987003        150.0  mastercard  117.0   debit  ...   NaN   NaN   NaN   NaN   
2987004        150.0  mastercard  102.0  credit  ...   0.0   0.0   0.0   0.0   

              V334 V335  V336  V337  V338  V339  
TransactionID                                    
2987000        NaN  NaN   NaN   NaN   NaN   NaN  
2987001        NaN  NaN   NaN   NaN   NaN   NaN  
2987002        NaN  NaN   NaN   NaN   NaN   NaN  
2987003        NaN  NaN   NaN   NaN   NaN   NaN  
2987004        0.0  0.0   0.0   0.0   0.0   0.0  

[5 rows x 393 columns]

In [4]:
# Info tabel train_transaction
train_transaction.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 590540 entries, 2987000 to 3577539
Columns: 393 entries, isFraud to V339
dtypes: float64(376), int64(3), object(14)
memory usage: 1.7+ GB


In [5]:
# Shape data train_transaction (Banyaknya baris dan kolom)
train_transaction.shape

(590540, 393)

In [6]:
# Nama kolom data train_transaction
train_transaction.columns

Index(['isFraud', 'TransactionDT', 'TransactionAmt', 'ProductCD', 'card1',
       'card2', 'card3', 'card4', 'card5', 'card6',
       ...
       'V330', 'V331', 'V332', 'V333', 'V334', 'V335', 'V336', 'V337', 'V338',
       'V339'],
      dtype='object', length=393)

In [7]:
# Menampilkan lima baris pertama data test_transaction
test_transaction.head()

TransactionDT  TransactionAmt ProductCD  card1  card2  card3  \
TransactionID                                                                 
3663549             18403224           31.95         W  10409  111.0  150.0   
3663550             18403263           49.00         W   4272  111.0  150.0   
3663551             18403310          171.00         W   4476  574.0  150.0   
3663552             18403310          284.95         W  10989  360.0  150.0   
3663553             18403317           67.95         W  18018  452.0  150.0   

                    card4  card5  card6  addr1  ...  V330  V331  V332 V333  \
TransactionID                                   ...                          
3663549              visa  226.0  debit  170.0  ...   NaN   NaN   NaN  NaN   
3663550              visa  226.0  debit  299.0  ...   NaN   NaN   NaN  NaN   
3663551              visa  226.0  debit  472.0  ...   NaN   NaN   NaN  NaN   
3663552              visa  166.0  debit  205.0  ...   NaN   NaN   NaN  NaN   
3663553        mastercard  117.0  debit  264.0  ...   NaN   NaN   NaN  NaN   

              V334  V335  V336  V337  V338  V339  
TransactionID                                     
3663549        NaN   NaN   NaN   NaN   NaN   NaN  
3663550        NaN   NaN   NaN   NaN   NaN   NaN  
3663551        NaN   NaN   NaN   NaN   NaN   NaN  
3663552        NaN   NaN   NaN   NaN   NaN   NaN  
3663553        NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 392 columns]

In [8]:
# Shape data test_transaction
test_transaction.shape

(506691, 392)

In [9]:
# Nama kolom data test_transaction
test_transaction.columns

Index(['TransactionDT', 'TransactionAmt', 'ProductCD', 'card1', 'card2',
       'card3', 'card4', 'card5', 'card6', 'addr1',
       ...
       'V330', 'V331', 'V332', 'V333', 'V334', 'V335', 'V336', 'V337', 'V338',
       'V339'],
      dtype='object', length=392)

In [10]:
# Menampilkan kolom target yaitu kolom isFraud
train_transaction.isFraud

TransactionID
2987000    0
2987001    0
2987002    0
2987003    0
2987004    0
          ..
3577535    0
3577536    0
3577537    0
3577538    0
3577539    0
Name: isFraud, Length: 590540, dtype: int64

In [11]:
# Cek Missing value pada kolom target
train_transaction.isFraud.isnull().sum()

0

In [12]:
# Persentasi transaksi penipuan pada data train_transaction
"percent of fraudulent train-transactions: ", len(train_transaction.loc[train_transaction.isFraud == 1])*100/len(train_transaction)

('percent of fraudulent train-transactions: ', 3.499000914417313)

Seperti yang bisa kita lihat, 3,5% pada data train adalah transaksi fraud/penipuan, ini merupakan imbalanced-dataset. Kami dapat berasumsi bahwa persentase transaksi fraud dalam data test memiliki besaran yang hampir sama.

Sekarang kita akan menghilangkan target dari dataframe train_transaction.

In [13]:
y_train = train_transaction["isFraud"]

# drop kolom target dari train dataframe
train_transaction = train_transaction.drop(columns = ['isFraud'])

print(y_train.shape, "\n")
print(train_transaction.shape, "\n")

(590540,) 

(590540, 392) 



Identity Data

In [14]:
# Menampilkan lima baris pertama data train_identity
train_identity.head()

id_01     id_02  id_03  id_04  id_05  id_06  id_07  id_08  \
TransactionID                                                              
2987004          0.0   70787.0    NaN    NaN    NaN    NaN    NaN    NaN   
2987008         -5.0   98945.0    NaN    NaN    0.0   -5.0    NaN    NaN   
2987010         -5.0  191631.0    0.0    0.0    0.0    0.0    NaN    NaN   
2987011         -5.0  221832.0    NaN    NaN    0.0   -6.0    NaN    NaN   
2987016          0.0    7460.0    0.0    0.0    1.0    0.0    NaN    NaN   

               id_09  id_10  ...                id_31 id_32      id_33  \
TransactionID                ...                                         
2987004          NaN    NaN  ...  samsung browser 6.2  32.0  2220x1080   
2987008          NaN    NaN  ...   mobile safari 11.0  32.0   1334x750   
2987010          0.0    0.0  ...          chrome 62.0   NaN        NaN   
2987011          NaN    NaN  ...          chrome 62.0   NaN        NaN   
2987016          0.0    0.0  ...          chrome 62.0  24.0   1280x800   

                        id_34 id_35 id_36  id_37  id_38  DeviceType  \
TransactionID                                                         
2987004        match_status:2     T     F      T      T      mobile   
2987008        match_status:1     T     F      F      T      mobile   
2987010                   NaN     F     F      T      T     desktop   
2987011                   NaN     F     F      T      T     desktop   
2987016        match_status:2     T     F      T      T     desktop   

                                  DeviceInfo  
TransactionID                                 
2987004        SAMSUNG SM-G892A Build/NRD90M  
2987008                           iOS Device  
2987010                              Windows  
2987011                                  NaN  
2987016                                MacOS  

[5 rows x 40 columns]

In [15]:
# Shape data train_identity
train_identity.shape

(144233, 40)

In [16]:
# Nama kolom data train_identity
train_identity.columns

Index(['id_01', 'id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_07', 'id_08',
       'id_09', 'id_10', 'id_11', 'id_12', 'id_13', 'id_14', 'id_15', 'id_16',
       'id_17', 'id_18', 'id_19', 'id_20', 'id_21', 'id_22', 'id_23', 'id_24',
       'id_25', 'id_26', 'id_27', 'id_28', 'id_29', 'id_30', 'id_31', 'id_32',
       'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType',
       'DeviceInfo'],
      dtype='object')

In [17]:
# Menampilkan lima baris pertama data test_identity
test_identity.head()

id-01     id-02  id-03  id-04  id-05  id-06  id-07  id-08  \
TransactionID                                                              
3663586        -45.0  280290.0    NaN    NaN    0.0    0.0    NaN    NaN   
3663588          0.0    3579.0    0.0    0.0    0.0    0.0    NaN    NaN   
3663597         -5.0  185210.0    NaN    NaN    1.0    0.0    NaN    NaN   
3663601        -45.0  252944.0    0.0    0.0    0.0    0.0    NaN    NaN   
3663602        -95.0  328680.0    NaN    NaN    7.0  -33.0    NaN    NaN   

               id-09  id-10  ...                    id-31 id-32     id-33  \
TransactionID                ...                                            
3663586          NaN    NaN  ...  chrome 67.0 for android   NaN       NaN   
3663588          0.0    0.0  ...  chrome 67.0 for android  24.0  1280x720   
3663597          NaN    NaN  ...       ie 11.0 for tablet   NaN       NaN   
3663601          0.0    0.0  ...  chrome 67.0 for android   NaN       NaN   
3663602          NaN    NaN  ...  chrome 67.0 for android   NaN       NaN   

                        id-34 id-35 id-36  id-37  id-38  DeviceType  \
TransactionID                                                         
3663586                   NaN     F     F      T      F      mobile   
3663588        match_status:2     T     F      T      T      mobile   
3663597                   NaN     F     T      T      F     desktop   
3663601                   NaN     F     F      T      F      mobile   
3663602                   NaN     F     F      T      F      mobile   

                                DeviceInfo  
TransactionID                               
3663586        MYA-L13 Build/HUAWEIMYA-L13  
3663588               LGLS676 Build/MXB48T  
3663597                        Trident/7.0  
3663601        MYA-L13 Build/HUAWEIMYA-L13  
3663602               SM-G9650 Build/R16NW  

[5 rows x 40 columns]

In [18]:
# Shape data test_identity
test_identity.shape

(141907, 40)

In [19]:
# Nama kolom data test_identity
test_identity.columns

Index(['id-01', 'id-02', 'id-03', 'id-04', 'id-05', 'id-06', 'id-07', 'id-08',
       'id-09', 'id-10', 'id-11', 'id-12', 'id-13', 'id-14', 'id-15', 'id-16',
       'id-17', 'id-18', 'id-19', 'id-20', 'id-21', 'id-22', 'id-23', 'id-24',
       'id-25', 'id-26', 'id-27', 'id-28', 'id-29', 'id-30', 'id-31', 'id-32',
       'id-33', 'id-34', 'id-35', 'id-36', 'id-37', 'id-38', 'DeviceType',
       'DeviceInfo'],
      dtype='object')

In [20]:
# Mengganti nama kolom pada data test_identity 
test_identity.columns = test_identity.columns.str.replace('-','_')
test_identity.columns

Index(['id_01', 'id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_07', 'id_08',
       'id_09', 'id_10', 'id_11', 'id_12', 'id_13', 'id_14', 'id_15', 'id_16',
       'id_17', 'id_18', 'id_19', 'id_20', 'id_21', 'id_22', 'id_23', 'id_24',
       'id_25', 'id_26', 'id_27', 'id_28', 'id_29', 'id_30', 'id_31', 'id_32',
       'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType',
       'DeviceInfo'],
      dtype='object')

In [21]:
# Menampilkan jumlah nilai berbeda pada fitur id (id_01)
train_identity.id_01.value_counts()

-5.0     82170
 0.0     19555
-10.0    11257
-20.0    11211
-15.0     5674
         ...  
-28.0        1
-72.0        1
-54.0        1
-89.0        1
-94.0        1
Name: id_01, Length: 77, dtype: int64

In [22]:
# Menampilkan jumlah nilai berbeda pada fitur id (id_07)
train_identity.id_07.value_counts()

 0.0     409
 16.0    245
 14.0    228
 12.0    208
 15.0    186
        ... 
-26.0      1
 51.0      1
-37.0      1
 46.0      1
 52.0      1
Name: id_07, Length: 84, dtype: int64

38 fitur id yang berbeda memiliki banyak nilai yang berbeda, tetapi 10 nilai unik pertama akan mencakup sekitar 90% dari semua data (estimasi dengan cepat)

In [23]:
# Menampilkan nilai yang berbeda pada fitur DeviceType
print(train_identity.DeviceType.value_counts(),'\n')


# Menampilkan nilai yang berbeda pada fitur DeviceInfo
print(train_identity.DeviceInfo.value_counts())

desktop    85165
mobile     55645
Name: DeviceType, dtype: int64 

Windows                  47722
iOS Device               19782
MacOS                    12573
Trident/7.0               7440
rv:11.0                   1901
                         ...  
IdeaTabA1000-F               1
SM-T580                      1
SM-G950U1 Build/R16NW        1
ZA990                        1
HTC6525LVW                   1
Name: DeviceInfo, Length: 1786, dtype: int64


-DeviceType hanya memiliki 2 nilai berbeda yaitu 'desktop' dan 'mobile'

-DeviceInfo memiliki sekitar 5 hingga 7 nilai berbeda yang mencakup sekitar 90% dari semua data  (diestimasi dengan cepat)

Sekarang kita akan menggabungkan data train dan data test sehingga kita memiliki satu kerangka data transaksi dan satu kerangka data identitas. Ini akan membuat prapemrosesan data kita jauh lebih mudah dibandingkan dengan melakukan semuanya dua kali pada kerangka data train dan test. Dan sebelum proses training model, kita akan membagi kerangka data menjadi data train dan data test .

In [24]:
# Concat train_transction dan tets transaction
transaction_data = pd.concat([train_transaction, test_transaction])
identity_data = pd.concat([train_identity, test_identity])

# setelah concatting
transaction_data.shape

(1097231, 392)

In [25]:
# Nama kolom data transaction
transaction_data.columns

Index(['TransactionDT', 'TransactionAmt', 'ProductCD', 'card1', 'card2',
       'card3', 'card4', 'card5', 'card6', 'addr1',
       ...
       'V330', 'V331', 'V332', 'V333', 'V334', 'V335', 'V336', 'V337', 'V338',
       'V339'],
      dtype='object', length=392)

In [26]:
#Index data transaction(label baris)
transaction_data.index

Int64Index([2987000, 2987001, 2987002, 2987003, 2987004, 2987005, 2987006,
            2987007, 2987008, 2987009,
            ...
            4170230, 4170231, 4170232, 4170233, 4170234, 4170235, 4170236,
            4170237, 4170238, 4170239],
           dtype='int64', name='TransactionID', length=1097231)

In [27]:
# Shape data identity 
identity_data.columns

Index(['id_01', 'id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_07', 'id_08',
       'id_09', 'id_10', 'id_11', 'id_12', 'id_13', 'id_14', 'id_15', 'id_16',
       'id_17', 'id_18', 'id_19', 'id_20', 'id_21', 'id_22', 'id_23', 'id_24',
       'id_25', 'id_26', 'id_27', 'id_28', 'id_29', 'id_30', 'id_31', 'id_32',
       'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType',
       'DeviceInfo'],
      dtype='object')

In [28]:
#Index data tidentity
identity_data.index

Int64Index([2987004, 2987008, 2987010, 2987011, 2987016, 2987017, 2987022,
            2987038, 2987040, 2987048,
            ...
            4170224, 4170225, 4170226, 4170228, 4170229, 4170230, 4170233,
            4170234, 4170236, 4170239],
           dtype='int64', name='TransactionID', length=286140)

Sekarang kita bisa menghapus train_transaction, train_identity, test_transaction, test_identity, karena kita tidak lagi membutuhkannya.

In [29]:
# sebelum menghapus dataframe, menyimpan shape dataframe dalam variabel
shape_of_train_trans = train_transaction.shape
shape_of_train_id    = train_identity.shape

shape_of_test_trans  = test_transaction.shape
shape_of_test_id     = test_identity.shape

# delete dataframe
del train_transaction
del train_identity
del test_transaction
del test_identity

print("deletion successful!")

deletion successful!


**Analisis Fitur**

In [30]:
# Membuat list fitur kategoris dan numerik dari identity_data

a = (identity_data.dtypes == 'object')
b = (identity_data.dtypes != 'object')
cat_id_cols = list(a[a].index)
num_id_cols = list(b[b].index) 

print(cat_id_cols, "\n")
print("number categorical identity features: ", len(cat_id_cols), "\n\n")
print(num_id_cols, "\n")
print("number numerical identity features: ", len(num_id_cols))

['id_12', 'id_15', 'id_16', 'id_23', 'id_27', 'id_28', 'id_29', 'id_30', 'id_31', 'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType', 'DeviceInfo'] 

number categorical identity features:  17 


['id_01', 'id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_07', 'id_08', 'id_09', 'id_10', 'id_11', 'id_13', 'id_14', 'id_17', 'id_18', 'id_19', 'id_20', 'id_21', 'id_22', 'id_24', 'id_25', 'id_26', 'id_32'] 

number numerical identity features:  23


In [31]:
# Membuat list fitur kategoris dan numerik dari transaction_data

c = (transaction_data.dtypes == 'object')
n = (transaction_data.dtypes != 'object')

# list fitur kategoris
cat_trans_cols = list(c[c].index)

# list fitur numerik
num_trans_cols = list(n[n].index) 

print(cat_trans_cols, "\n")
print("number categorical transaction features: ", len(cat_trans_cols), "\n\n")
print(num_trans_cols, "\n")
print("number numerical transaction features: ", len(num_trans_cols))

['ProductCD', 'card4', 'card6', 'P_emaildomain', 'R_emaildomain', 'M1', 'M2', 'M3', 'M4', 'M5', 'M6', 'M7', 'M8', 'M9'] 

number categorical transaction features:  14 


['TransactionDT', 'TransactionAmt', 'card1', 'card2', 'card3', 'card5', 'addr1', 'addr2', 'dist1', 'dist2', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11', 'C12', 'C13', 'C14', 'D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9', 'D10', 'D11', 'D12', 'D13', 'D14', 'D15', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31', 'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61', 'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V69', 'V70', 'V71', 'V72', 'V73', 'V74', 'V75', 'V76', 'V77', 'V78', 

Analisis data Identity

In [32]:
# Menghitung Missing Value categorical identity feature
for i in cat_id_cols:
    print(identity_data[i].value_counts())
    print(i, "missing values: ", identity_data[i].isnull().sum())
    print(identity_data[i].isnull().sum()*100/ len(identity_data[i]), "\n")

NotFound    243920
Found        42220
Name: id_12, dtype: int64
id_12 missing values:  0
0.0 

Found      135690
New        119397
Unknown     22875
Name: id_15, dtype: int64
id_15 missing values:  8178
2.8580415181379744 

Found       132805
NotFound    122282
Name: id_16, dtype: int64
id_16 missing values:  31053
10.852379953868736 

IP_PROXY:TRANSPARENT    7203
IP_PROXY:ANONYMOUS      2010
IP_PROXY:HIDDEN         1018
Name: id_23, dtype: int64
id_23 missing values:  275909
96.42447752848256 

Found       10214
NotFound       17
Name: id_27, dtype: int64
id_27 missing values:  275909
96.42447752848256 

Found    151813
New      125943
Name: id_28, dtype: int64
id_28 missing values:  8384
2.930034248969036 

Found       149264
NotFound    128492
Name: id_29, dtype: int64
id_29 missing values:  8384
2.930034248969036 

Windows 10          42170
Windows 7           23478
iOS 12.1.0           6349
Mac OS X 10_12_6     3884
iOS 11.2.1           3824
                    ...  
Mac OS X 10_1

In [33]:
# Menghitung persentase missing value pada categorical identity faeture
low_missing_cat_id_cols = []      # lower than 15% missing values
medium_missing_cat_id_cols = []   # between 15% and 60% missing
many_missing_cat_id_cols = []     # more than 60% missing

for i in cat_id_cols:
    percentage = identity_data[i].isnull().sum() * 100 / len(identity_data[i])
    if percentage < 15:
        low_missing_cat_id_cols.append(i)
    elif percentage >= 15 and percentage < 60:
        medium_missing_cat_id_cols.append(i)
    else:
        many_missing_cat_id_cols.append(i)
        
print("cat_id_cols: \n\n")      
print("number low missing: ", len(low_missing_cat_id_cols), "\n")
print("number medium missing: ", len(medium_missing_cat_id_cols), "\n")
print("number many missing: ", len(many_missing_cat_id_cols), "\n")

cat_id_cols: 


number low missing:  11 

number medium missing:  4 

number many missing:  2 



In [34]:
# Menghitung persentase missing value numerical identity features
low_missing_num_id_cols = []      # lower than 15% missing values
medium_missing_num_id_cols = []   # between 15% and 60% missing
many_missing_num_id_cols = []     # more than 60% missing

for i in num_id_cols:
    percentage = identity_data[i].isnull().sum() * 100 / len(identity_data[i])
    if percentage < 15:
        low_missing_num_id_cols.append(i)
    elif percentage >= 15 and percentage < 60:
        medium_missing_num_id_cols.append(i)
    else:
        many_missing_num_id_cols.append(i)
        
print("num_id_cols: \n\n")        
print("number low missing: ", len(low_missing_num_id_cols), "\n")
print("number medium missing: ", len(medium_missing_num_id_cols), "\n")
print("number many missing: ", len(many_missing_num_id_cols), "\n")

num_id_cols: 


number low missing:  9 

number medium missing:  6 

number many missing:  8 



Analisis data transaksi

In [35]:
# Menghitung missing value categorical transaction features
low_missing_cat_trans_cols = []      # lower than 15% missing values
medium_missing_cat_trans_cols = []   # between 15% and 60% missing
many_missing_cat_trans_cols = []     # more than 60% missing

for i in cat_trans_cols:
    percentage = transaction_data[i].isnull().sum() * 100 / len(transaction_data[i])
    if percentage < 15:
        low_missing_cat_trans_cols.append(i)
    elif percentage >= 15 and percentage < 60:
        medium_missing_cat_trans_cols.append(i)
    else:
        many_missing_cat_trans_cols.append(i)
        
print("cat_trans_cols: \n\n")        
print("number low missing: ", len(low_missing_cat_trans_cols), "\n")
print("number medium missing: ", len(medium_missing_cat_trans_cols), "\n")
print("number many missing: ", len(many_missing_cat_trans_cols), "\n")

cat_trans_cols: 


number low missing:  4 

number medium missing:  8 

number many missing:  2 



In [36]:
# Missing value numerical transaction features
low_missing_num_trans_cols = []      # lower than 15% missing values
medium_missing_num_trans_cols = []   # between 15% and 60% missing
many_missing_num_trans_cols = []     # more than 60% missing

for i in num_trans_cols:
    percentage = transaction_data[i].isnull().sum() * 100 / len(transaction_data[i])
    if percentage < 15:
        low_missing_num_trans_cols.append(i)
    elif percentage >= 15 and percentage < 60:
        medium_missing_num_trans_cols.append(i)
    else:
        many_missing_num_trans_cols.append(i)
        
print("num_trans_cols: \n\n")        
print("number low missing: ", len(low_missing_num_trans_cols), "\n")
print("number medium missing: ", len(medium_missing_num_trans_cols), "\n")
print("number many missing: ", len(many_missing_num_trans_cols), "\n")

num_trans_cols: 


number low missing:  176 

number medium missing:  35 

number many missing:  167 



Kesimpulan:

kita memiliki 2 dataframes: transaction_data and identity_data

features:

transaction_data: 14 categorical and 378 numerical features

identity_data: 17 categorical and 23 numerical features

missing values:

cat_trans_cols: 4 low, 8 medium, 2 many num_trans_cols: 176 low, 35 medium, 167 many

cat_id_cols: 11 low, 4 medium, 2 many num_id_cols: 9 low, 6 medium, 8 many

kita akan membuang semua fitur numerik dengan many missing value

mengganti dengan 'median' semua fitur numerik medium missing value

dan mengganti dengan 'mean' semua fitur numerik dengan medium missing value.

****Analisis Fitur Numerik****

Pertama-tama mari kita menghapus semua fitur numerik dengan many missing value.

In [37]:
#  Dropping fitur dengan many missing values
print("shape before dropping num_trans_cols: ", transaction_data.shape, "\n")        
transaction_data = transaction_data.drop(columns = many_missing_num_trans_cols)
print("shape after dropping num_trans_cols: ", transaction_data.shape, "\n\n")    


print("shape before dropping num_id_cols: ", identity_data.shape, "\n")        
identity_data = identity_data.drop(columns = many_missing_num_id_cols)
print("shape after dropping num_id_cols: ", identity_data.shape, "\n")

# Karena kita telah menghapus beberapa kolom numerik dari dataframe
# kita harus membuat kembali list 'num_trans_col' dan 'num_id cols'
# agar kolom yang telah di drop  tidak termasuk lagi didalamnya
n = (transaction_data.dtypes != 'object')
num_trans_cols = list(n[n].index) 

n = (identity_data.dtypes != 'object')
num_id_cols = list(n[n].index) 

shape before dropping num_trans_cols:  (1097231, 392) 

shape after dropping num_trans_cols:  (1097231, 225) 


shape before dropping num_id_cols:  (286140, 40) 

shape after dropping num_id_cols:  (286140, 32) 



Selanjutnya impute fitur dengan low missing value dengan mean

In [38]:
#  Impute fitur low missing values deangan mean
# 1.  Impute the transaction features
from sklearn.impute import SimpleImputer


my_imputer = SimpleImputer(strategy = 'mean') 
my_imputer.fit(transaction_data[low_missing_num_trans_cols])

transaction_data[low_missing_num_trans_cols] = my_imputer.transform(transaction_data[low_missing_num_trans_cols])


In [39]:
# 2.Impute the identity features

my_imputer = SimpleImputer(strategy = 'mean') 
my_imputer.fit(identity_data[low_missing_num_id_cols])

identity_data[low_missing_num_id_cols] = my_imputer.transform(identity_data[low_missing_num_id_cols])


lalu, Impute fitur medium missing values dengan median.

In [40]:
# 1 Impute fitur data_transaction 

my_imputer = SimpleImputer(strategy = 'median') 
my_imputer.fit(transaction_data[medium_missing_num_trans_cols])

transaction_data[medium_missing_num_trans_cols] = my_imputer.transform(transaction_data[medium_missing_num_trans_cols])


In [41]:
#2 Impute fitur data_identity

my_imputer = SimpleImputer(strategy = 'median') 
my_imputer.fit(identity_data[medium_missing_num_id_cols])

identity_data[medium_missing_num_id_cols] = my_imputer.transform(identity_data[medium_missing_num_id_cols])


In [42]:
#Final check jika masih terdapat missing values in num. columns
print(transaction_data[num_trans_cols].isnull().sum().sum())
print(identity_data[num_id_cols].isnull().sum().sum())

0
0


***Analisis Fitur Kategorik***

Untuk data dengan fitur kategorik kita akan memproses dataframe train_transaction dan train_identity yaitu dengan cara:

1.Drop fitur untuk many missing value

2.Label-encode untuk fitur dengan kardinalitas tinggi

3.Onehot-Encode untuk fitur dengan kardinalitas rendah

Drop fitur kategorik dengan many missing values

In [43]:
#1 Drop fitur kategorik dengan many missing values
print("shape before dropping many_missing_cat_trans_cols: ", transaction_data.shape, "\n")        
transaction_data = transaction_data.drop(columns = many_missing_cat_trans_cols)
print("shape after dropping many_missing_cat_trans_cols: ", transaction_data.shape, "\n\n")    

print("shape before dropping many_missing_cat_id_cols: ", identity_data.shape, "\n")        
identity_data = identity_data.drop(columns = many_missing_cat_id_cols)
print("shape after dropping many_missing_cat_id_cols: ", identity_data.shape, "\n")

# karena kita telah menghapus/drop beberapa kolom dari dataframe
# kita akan membuat list 'cat_trans_cols' dan 'cat_id_cols' 
# agar kolom yang telah didrop tidak termasuk lagi didalamnya

c = (transaction_data.dtypes == 'object')
cat_trans_cols = list(c[c].index) 

c = (identity_data.dtypes == 'object')
cat_id_cols = list(c[c].index) 

shape before dropping many_missing_cat_trans_cols:  (1097231, 225) 

shape after dropping many_missing_cat_trans_cols:  (1097231, 223) 


shape before dropping many_missing_cat_id_cols:  (286140, 32) 

shape after dropping many_missing_cat_id_cols:  (286140, 30) 



Label-Encode untuk features dengan kardinalitas tinggi

In [44]:
# 2 Label-Encode untuk features dengan kardinalitas tinggi

# 2.1 Pertama kita hitung kardinalitas dari fitur kategorical untuk transaction data
for col in cat_trans_cols:
    print(col, transaction_data[col].nunique(), "\n")

ProductCD 5 

card4 4 

card6 4 

P_emaildomain 60 

M1 2 

M2 2 

M3 2 

M4 3 

M6 2 

M7 2 

M8 2 

M9 2 



Seperti yang bisa kita lihat ada dua kelompok kardinalitas: kardinalitas rendah dengan nilai unik (2,3,4,5), dan kardinalitas tinggi komparatif dengan 60 nilai unik.

In [45]:
# Membuat list untuk membagi fitur atas kardinalitas rendah dan kardinalitas tinggi pada data transaksi
low_card_trans_cols = ["ProductCD", "card4", "card6", "M1", "M2", "M3", "M4", "M6", "M7", "M8", "M9"]
high_card_trans_cols = ["P_emaildomain"]

print("lists successfully created!")

lists successfully created!


Sebelum kita melakukan label-encode, kita akan mengganti semua NaN dengan nilai yang paling sering muncul pada setiap kolom.

In [46]:
# Replace nilai NaN dengan nilai yang sering muncul pada data transaction
for i in cat_trans_cols:
    most_frequent_value = transaction_data[i].mode()[0]
    print("For column: ", i, "the most frequent value is: ", most_frequent_value, "\n")
    transaction_data[i].fillna(most_frequent_value, inplace = True)

For column:  ProductCD the most frequent value is:  W 

For column:  card4 the most frequent value is:  visa 

For column:  card6 the most frequent value is:  debit 

For column:  P_emaildomain the most frequent value is:  gmail.com 

For column:  M1 the most frequent value is:  T 

For column:  M2 the most frequent value is:  T 

For column:  M3 the most frequent value is:  T 

For column:  M4 the most frequent value is:  M0 

For column:  M6 the most frequent value is:  F 

For column:  M7 the most frequent value is:  F 

For column:  M8 the most frequent value is:  F 

For column:  M9 the most frequent value is:  T 



In [47]:
# Melakukan label-encode
from sklearn.preprocessing import LabelEncoder
    
label_encoder = LabelEncoder()
print("transaction_data.shape before label-encoding: ", transaction_data.shape, "\n")

transaction_data[high_card_trans_cols] = label_encoder.fit_transform(transaction_data[high_card_trans_cols])

print("transaction_data.shape after label-encoding: ", transaction_data.shape, "\n")
print("transaction_data[high_card_trans_cols] after label_encoding: ",transaction_data[high_card_trans_cols], "\n")

transaction_data.shape before label-encoding:  (1097231, 223) 



/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


transaction_data.shape after label-encoding:  (1097231, 223) 

transaction_data[high_card_trans_cols] after label_encoding:                 P_emaildomain
TransactionID               
2987000                   16
2987001                   16
2987002                   35
2987003                   54
2987004                   16
...                      ...
4170235                   16
4170236                   19
4170237                   19
4170238                   19
4170239                   19

[1097231 rows x 1 columns] 



In [48]:
# 2.2 Menghitung kardinalitas dari fitur kategorical untuk identity data
for col in cat_id_cols:
    print(col, identity_data[col].nunique(), "\n")

id_12 2 

id_15 3 

id_16 2 

id_28 2 

id_29 2 

id_30 87 

id_31 172 

id_33 461 

id_34 4 

id_35 2 

id_36 2 

id_37 2 

id_38 2 

DeviceType 2 

DeviceInfo 2799 



In [49]:
# Membuat list untuk membagi fitur atas kardinalitas rendah dan kardinalitas tinggi pada data identity
low_card_id_cols =  ["id_12", "id_15", "id_16", "id_28", "id_29", "id_34", "id_35", "id_36", "id_37", "id_38", "DeviceType"]
high_card_id_cols = ["id_30", "id_31", "id_33", "DeviceInfo"]
    
print("lists successfully created!")

lists successfully created!


In [50]:
# Replace nilai NaN dengan nilai yang sering muncul pada data transaction
for i in cat_id_cols:
    most_frequent_value = identity_data[i].mode()[0]
    print("For column: ", i, "the most frequent value is: ", most_frequent_value, "\n")
    identity_data[i].fillna(most_frequent_value, inplace = True)

For column:  id_12 the most frequent value is:  NotFound 

For column:  id_15 the most frequent value is:  Found 

For column:  id_16 the most frequent value is:  Found 

For column:  id_28 the most frequent value is:  Found 

For column:  id_29 the most frequent value is:  Found 

For column:  id_30 the most frequent value is:  Windows 10 

For column:  id_31 the most frequent value is:  mobile safari 11.0 

For column:  id_33 the most frequent value is:  1920x1080 

For column:  id_34 the most frequent value is:  match_status:2 

For column:  id_35 the most frequent value is:  T 

For column:  id_36 the most frequent value is:  F 

For column:  id_37 the most frequent value is:  T 

For column:  id_38 the most frequent value is:  F 

For column:  DeviceType the most frequent value is:  desktop 

For column:  DeviceInfo the most frequent value is:  Windows 



In [51]:
# Melakukan label-encode
label_encoder = LabelEncoder()

print("identity_data.shape before label-encoding: ", identity_data.shape, "\n")

for col in high_card_id_cols:
    identity_data[col] = label_encoder.fit_transform(identity_data[col])

print("identity_data.shape after label-encoding: ", identity_data.shape, "\n")
print("identity_data[high_card_id_cols] after label_encoding: ",identity_data[high_card_id_cols], "\n")

identity_data.shape before label-encoding:  (286140, 30) 

identity_data.shape after label-encoding:  (286140, 30) 

identity_data[high_card_id_cols] after label_encoding:                 id_30  id_31  id_33  DeviceInfo
TransactionID                                 
2987004            7    161    268        1565
2987008           70    130     80        2693
2987010           49     46    216        2526
2987011           49     46    216        2526
2987016           26     46     68        1170
...              ...    ...    ...         ...
4170230           49     74    216        2165
4170233           49     74    216        2106
4170234           63    129     80        2693
4170236           49     22    216         141
4170239           49    166    216        1529

[286140 rows x 4 columns] 



Onehot-Encode untuk features dengan kardinalitas rendah

Kita akan melakukan onehot-encode pada fitur kardinalitas rendah. karena proses onehot-encoder akan menghasilkan lebih banyak kolom, nama kolom, nilai kolom dan posisi kolom semuanya akan dihasilkan secara otomatis. Oleh karena itu kita harus membuat kerangka data terpisah untuk setiap proses onehot-encode, dan kemudian menyatukan kembali kerangka data ini setelahnya. 

In [52]:
# Melakukan onehot-encoding pada data transaction
print("shape before encoding: ", transaction_data.shape, "\n")
print("columns to encode: ", low_card_trans_cols, "\n")
print("transaction_data.columns.to_list() before encoding: ", transaction_data.columns.to_list(), "\n")


# onehot encoding
low_card_trans_encoded = pd.get_dummies(transaction_data[low_card_trans_cols], dummy_na = False)
transaction_data.drop(columns = low_card_trans_cols, inplace = True)

print("shape after encoding: ", transaction_data.shape, "\n\n")
print("shape of new dataframe: ", low_card_trans_encoded.shape, "\n\n")
print("newly generated columns: ", low_card_trans_encoded.columns, "\n")
print("low_card_trans_encoded.info(): ", low_card_trans_encoded.info(),"\n")
print("transaction_data.columns.to_list() after encoding: ", transaction_data.columns.to_list(), "\n")

shape before encoding:  (1097231, 223) 

columns to encode:  ['ProductCD', 'card4', 'card6', 'M1', 'M2', 'M3', 'M4', 'M6', 'M7', 'M8', 'M9'] 

transaction_data.columns.to_list() before encoding:  ['TransactionDT', 'TransactionAmt', 'ProductCD', 'card1', 'card2', 'card3', 'card4', 'card5', 'card6', 'addr1', 'addr2', 'dist1', 'P_emaildomain', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11', 'C12', 'C13', 'C14', 'D1', 'D2', 'D3', 'D4', 'D5', 'D10', 'D11', 'D15', 'M1', 'M2', 'M3', 'M4', 'M6', 'M7', 'M8', 'M9', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31', 'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61', 'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V

Sekarang kita lakukan hal yang sama untuk fitur kategoris dari data identitas:

In [53]:
# Melakukan onehot-encoding pada data identity
print("shape before encoding: ", identity_data.shape, "\n")
print("columns to encode: ", low_card_id_cols, "\n")

# onehot encoding
low_card_id_encoded = pd.get_dummies(identity_data[low_card_id_cols], dummy_na = False)
identity_data.drop(columns = low_card_id_cols, inplace = True)


print("shape after encoding: ", identity_data.shape, "\n\n")
print("shape of new dataframe: ", low_card_id_encoded.shape, "\n\n")
print("newly generated columns: ", low_card_id_encoded.columns, "\n")
print("low_card_id_encoded.info(): ", low_card_id_encoded.info())

shape before encoding:  (286140, 30) 

columns to encode:  ['id_12', 'id_15', 'id_16', 'id_28', 'id_29', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType'] 

shape after encoding:  (286140, 19) 


shape of new dataframe:  (286140, 25) 


newly generated columns:  Index(['id_12_Found', 'id_12_NotFound', 'id_15_Found', 'id_15_New',
       'id_15_Unknown', 'id_16_Found', 'id_16_NotFound', 'id_28_Found',
       'id_28_New', 'id_29_Found', 'id_29_NotFound', 'id_34_match_status:-1',
       'id_34_match_status:0', 'id_34_match_status:1', 'id_34_match_status:2',
       'id_35_F', 'id_35_T', 'id_36_F', 'id_36_T', 'id_37_F', 'id_37_T',
       'id_38_F', 'id_38_T', 'DeviceType_desktop', 'DeviceType_mobile'],
      dtype='object') 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 286140 entries, 2987004 to 4170239
Data columns (total 25 columns):
 #   Column                 Non-Null Count   Dtype
---  ------                 --------------   -----
 0   id_12_Found            286140 non-nu

**Terakhir cek missing value**

In [54]:
# Cek missing value pada data transaction
print(transaction_data.isnull().sum().sum(), "\n")
print(low_card_trans_encoded.isnull().sum().sum())

0 

0


In [55]:
# Cek missing value pada data identity
print(identity_data.isnull().sum().sum(), "\n")
print(low_card_id_encoded.isnull().sum().sum())

0 

0


In [56]:
# Info data encoded
print(transaction_data.info(), "\n")
print(low_card_trans_encoded.info())

print(identity_data.info(), "\n")
print(low_card_id_encoded.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1097231 entries, 2987000 to 4170239
Columns: 212 entries, TransactionDT to V321
dtypes: float64(211), int64(1)
memory usage: 1.7 GB
None 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1097231 entries, 2987000 to 4170239
Data columns (total 30 columns):
 #   Column                  Non-Null Count    Dtype
---  ------                  --------------    -----
 0   ProductCD_C             1097231 non-null  uint8
 1   ProductCD_H             1097231 non-null  uint8
 2   ProductCD_R             1097231 non-null  uint8
 3   ProductCD_S             1097231 non-null  uint8
 4   ProductCD_W             1097231 non-null  uint8
 5   card4_american express  1097231 non-null  uint8
 6   card4_discover          1097231 non-null  uint8
 7   card4_mastercard        1097231 non-null  uint8
 8   card4_visa              1097231 non-null  uint8
 9   card6_charge card       1097231 non-null  uint8
 10  card6_credit            1097231 non-null  uint8
 11

**Concat data encoded**

Saat ini kita memiliki 2 dataframe asli transaction_data dan identity_data yang berisi fitur kategorikal yang telah dilakukan label-encode , dan 2 dataframe onehot-encoded yaitu
low_card_trans_encoded, low_card_id_encoded.
Sekarang kita akan menggabungkan dataframe ini menjadi satu dataframe besar, lalu kita harus membagi transaction_data menjadi train_transaction dan test_transaction, lalu melakukan hal yang sama dengan identity_data.

In [57]:
# Concat transaction_data dan low_card_trans_encoded
print("transaction_data.shape before concatting: ", transaction_data.shape, "\n")
print("low_card_trans_encoded.shape before concatting: ", low_card_trans_encoded.shape, "\n")

transaction_concatted = pd.concat([transaction_data, low_card_trans_encoded], axis = 1)

print("transaction_concatted.shape after concatting: ", transaction_concatted.shape, "\n")
print("transaction_concatted.columns after concatting: ", transaction_concatted.columns, "\n")

#del low_card_trans_encoded
#del transaction_data

print(transaction_concatted.info())

transaction_data.shape before concatting:  (1097231, 212) 

low_card_trans_encoded.shape before concatting:  (1097231, 30) 

transaction_concatted.shape after concatting:  (1097231, 242) 

transaction_concatted.columns after concatting:  Index(['TransactionDT', 'TransactionAmt', 'card1', 'card2', 'card3', 'card5',
       'addr1', 'addr2', 'dist1', 'P_emaildomain',
       ...
       'M4_M1', 'M4_M2', 'M6_F', 'M6_T', 'M7_F', 'M7_T', 'M8_F', 'M8_T',
       'M9_F', 'M9_T'],
      dtype='object', length=242) 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1097231 entries, 2987000 to 4170239
Columns: 242 entries, TransactionDT to M9_T
dtypes: float64(211), int64(1), uint8(30)
memory usage: 1.8 GB
None


In [58]:
# Concat identity_data dan low_card_id_encoded
print("identity_data.shape before concatting: ", identity_data.shape, "\n")
print("low_card_id_encoded.shape before concatting: ", low_card_id_encoded.shape, "\n")

identity_concatted = pd.concat([identity_data, low_card_id_encoded], axis = 1)

print("identity_concatted.shape after concatting: ", identity_concatted.shape, "\n")
print("identity_concatted.columns after concatting: ", identity_concatted.columns, "\n")

#del low_card_id_encoded
#del identity_data

print(identity_concatted.info())

identity_data.shape before concatting:  (286140, 19) 

low_card_id_encoded.shape before concatting:  (286140, 25) 

identity_concatted.shape after concatting:  (286140, 44) 

identity_concatted.columns after concatting:  Index(['id_01', 'id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_09', 'id_10',
       'id_11', 'id_13', 'id_14', 'id_17', 'id_19', 'id_20', 'id_30', 'id_31',
       'id_32', 'id_33', 'DeviceInfo', 'id_12_Found', 'id_12_NotFound',
       'id_15_Found', 'id_15_New', 'id_15_Unknown', 'id_16_Found',
       'id_16_NotFound', 'id_28_Found', 'id_28_New', 'id_29_Found',
       'id_29_NotFound', 'id_34_match_status:-1', 'id_34_match_status:0',
       'id_34_match_status:1', 'id_34_match_status:2', 'id_35_F', 'id_35_T',
       'id_36_F', 'id_36_T', 'id_37_F', 'id_37_T', 'id_38_F', 'id_38_T',
       'DeviceType_desktop', 'DeviceType_mobile'],
      dtype='object') 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 286140 entries, 2987004 to 4170239
Data columns (total 44 columns

Sekarang kita memiliki 2 dataframe yaitu transaction_concatted dan identity_concatted, yang berisi semua fitur yang akan kita gunakan, semua fitur kategorikal telah dikodekan dan semua nilai yang hilang telah dihapus.

Sebelum kita memulai proses training model, kita akan membagi transaction_concatted menjadi train_transaction dan test_transaction, dan kita juga akan membagi identity_concatted kita.

Untuk proses pemisahan kita akan membagi kerangka data menjadi dua bagian,namun karena perbedaan shape transaction_concatted (1097231, 253) dan identity_concatted (286140, 55), kita tidak dapat menggabungkan terlebih dahulu dan kemudian memisahkannya.

Ini tidak akan menghasilkan 4 dataframe train_transaction, test_transaction, train_identity, test_identity yang benar, yang akan kita gunakan. Untuk mendapatkan kerangka data train dan test yang benar, kita harus memisahkan terlebih dahulu, lalu menggabungkannya.

Karena shape yang berbeda dari kedua kerangka data, maka akan ada sejumlah besar NaN yang dihasilkan. Hal ini disebabkan oleh fakta bahwa dataframe identity_concatted tidak memiliki kolom yang sama dengan transaction_concatted. Karenanya semua nilai kosong yang dihasilkan oleh proses penggabungan akan diisi dengan NaN. Kita tidak akan memperhitungkan nilai-nilai yang hilang ini, kita hanya akan memberi tahu model untuk mengabaikan nilai-nilai ini.

In [59]:
# Membagi/ memisahkan data transaction_concated menjadi train_transaction dan test_transaction
print("transaction_concatted.shape before splitting up: ", transaction_concatted.shape, "\n")

# shape dari train_transaction : (590540, 393), 
# shape dari test_transaction : (506691, 392)
train_transaction = transaction_concatted.iloc[0:590540]
test_transaction = transaction_concatted.iloc[590540:]

print("train_transaction.shape after splitting up: ", train_transaction.shape, "\n")
print("test_transaction.shape after splitting up: ", test_transaction.shape, "\n")

transaction_concatted.shape before splitting up:  (1097231, 242) 

train_transaction.shape after splitting up:  (590540, 242) 

test_transaction.shape after splitting up:  (506691, 242) 



In [60]:
# Membagi/ memisahkan data identity_concated menjadi train_identity dan test_identity
print("identity_concatted.shape before splitting up: ", identity_concatted.shape, "\n")

# shape dari train_identity : (144233, 40)
# shape dari test_identity : (141907, 40)
train_identity = identity_concatted.iloc[0:144233]
test_identity = identity_concatted.iloc[144233:]

print("train_identity.shape after splitting up: ", train_identity.shape, "\n")
print("test_identity.shape after splitting up: ", test_identity.shape, "\n")

identity_concatted.shape before splitting up:  (286140, 44) 

train_identity.shape after splitting up:  (144233, 44) 

test_identity.shape after splitting up:  (141907, 44) 



Selanjutnya menggabungkan data untuk mendapatkan dua kerangka data yaitu data train dan data test

In [61]:
# Data Train

print("train_transaction.shape before concatting: ", train_transaction.shape, "\n")
print("train_identity.shape before concatting: ", train_identity.shape, "\n")

# Concat train_transaction dan train_identity
train_data  = pd.concat([train_transaction, train_identity], axis = 1)

print("train_data.shape: ", train_data.shape)

train_transaction.shape before concatting:  (590540, 242) 

train_identity.shape before concatting:  (144233, 44) 

train_data.shape:  (590540, 286)


In [62]:
# Cek missing value pada data train
counter = 0

for i in train_data.columns:
    
    summ = train_data[i].isnull().sum()
    print(i, summ)
    if summ > 0:
        counter += 1
        
print("\n number of columns with missing values: ", counter)

TransactionDT 0
TransactionAmt 0
card1 0
card2 0
card3 0
card5 0
addr1 0
addr2 0
dist1 0
P_emaildomain 0
C1 0
C2 0
C3 0
C4 0
C5 0
C6 0
C7 0
C8 0
C9 0
C10 0
C11 0
C12 0
C13 0
C14 0
D1 0
D2 0
D3 0
D4 0
D5 0
D10 0
D11 0
D15 0
V1 0
V2 0
V3 0
V4 0
V5 0
V6 0
V7 0
V8 0
V9 0
V10 0
V11 0
V12 0
V13 0
V14 0
V15 0
V16 0
V17 0
V18 0
V19 0
V20 0
V21 0
V22 0
V23 0
V24 0
V25 0
V26 0
V27 0
V28 0
V29 0
V30 0
V31 0
V32 0
V33 0
V34 0
V35 0
V36 0
V37 0
V38 0
V39 0
V40 0
V41 0
V42 0
V43 0
V44 0
V45 0
V46 0
V47 0
V48 0
V49 0
V50 0
V51 0
V52 0
V53 0
V54 0
V55 0
V56 0
V57 0
V58 0
V59 0
V60 0
V61 0
V62 0
V63 0
V64 0
V65 0
V66 0
V67 0
V68 0
V69 0
V70 0
V71 0
V72 0
V73 0
V74 0
V75 0
V76 0
V77 0
V78 0
V79 0
V80 0
V81 0
V82 0
V83 0
V84 0
V85 0
V86 0
V87 0
V88 0
V89 0
V90 0
V91 0
V92 0
V93 0
V94 0
V95 0
V96 0
V97 0
V98 0
V99 0
V100 0
V101 0
V102 0
V103 0
V104 0
V105 0
V106 0
V107 0
V108 0
V109 0
V110 0
V111 0
V112 0
V113 0
V114 0
V115 0
V116 0
V117 0
V118 0
V119 0
V120 0
V121 0
V122 0
V123 0
V124 0
V125 0
V126 0
V12

In [63]:
# Data test
print("test_transaction.shape before concatting: ", test_transaction.shape, "\n")
print("test_identity.shape before concatting: ", test_identity.shape, "\n")

# Concat test_transaction dan testb identity
test_data  = pd.concat([test_transaction, test_identity], axis = 1)

print("test_data.shape: ", test_data.shape)

test_transaction.shape before concatting:  (506691, 242) 

test_identity.shape before concatting:  (141907, 44) 

test_data.shape:  (506691, 286)


In [64]:
# cek missing value pada data test
counter = 0

for i in test_data.columns:
    
    summ = test_data[i].isnull().sum()
    print(i, summ)
    if summ > 0:
        counter += 1
        
print("\n number of columns with missing values: ", counter)

TransactionDT 0
TransactionAmt 0
card1 0
card2 0
card3 0
card5 0
addr1 0
addr2 0
dist1 0
P_emaildomain 0
C1 0
C2 0
C3 0
C4 0
C5 0
C6 0
C7 0
C8 0
C9 0
C10 0
C11 0
C12 0
C13 0
C14 0
D1 0
D2 0
D3 0
D4 0
D5 0
D10 0
D11 0
D15 0
V1 0
V2 0
V3 0
V4 0
V5 0
V6 0
V7 0
V8 0
V9 0
V10 0
V11 0
V12 0
V13 0
V14 0
V15 0
V16 0
V17 0
V18 0
V19 0
V20 0
V21 0
V22 0
V23 0
V24 0
V25 0
V26 0
V27 0
V28 0
V29 0
V30 0
V31 0
V32 0
V33 0
V34 0
V35 0
V36 0
V37 0
V38 0
V39 0
V40 0
V41 0
V42 0
V43 0
V44 0
V45 0
V46 0
V47 0
V48 0
V49 0
V50 0
V51 0
V52 0
V53 0
V54 0
V55 0
V56 0
V57 0
V58 0
V59 0
V60 0
V61 0
V62 0
V63 0
V64 0
V65 0
V66 0
V67 0
V68 0
V69 0
V70 0
V71 0
V72 0
V73 0
V74 0
V75 0
V76 0
V77 0
V78 0
V79 0
V80 0
V81 0
V82 0
V83 0
V84 0
V85 0
V86 0
V87 0
V88 0
V89 0
V90 0
V91 0
V92 0
V93 0
V94 0
V95 0
V96 0
V97 0
V98 0
V99 0
V100 0
V101 0
V102 0
V103 0
V104 0
V105 0
V106 0
V107 0
V108 0
V109 0
V110 0
V111 0
V112 0
V113 0
V114 0
V115 0
V116 0
V117 0
V118 0
V119 0
V120 0
V121 0
V122 0
V123 0
V124 0
V125 0
V126 0
V12

Mari kita lihat kolom acak mana saja yang berisi banyak NaN ini:

In [65]:
# Menampilkan kolom  id_35_F
print(test_data["id_35_F"])

TransactionID
3663549    NaN
3663550    NaN
3663551    NaN
3663552    NaN
3663553    NaN
          ... 
4170235    NaN
4170236    1.0
4170237    NaN
4170238    NaN
4170239    1.0
Name: id_35_F, Length: 506691, dtype: float64


Hal baiknya adalah kita tidak perlu menghapus NaN ini, kita cukup membiarkan model xgb kita menanganinya.

**Train the model**

Kami memilih XGBClassifier sebagai model kami dan memulai proses pelatihan.

In [66]:
from xgboost import XGBClassifier


clf = XGBClassifier(objective = 'binary:logistic',
                    gamma = 0.05,
                    colsample_bytree = 0.5, 
                    eval_metric = 'auc',
                    n_estimators = 1350,         
                    max_depth = 8,
                    min_child_weight = 2, 
                    learning_rate = 0.02,
                    subsample = 0.8,
                    n_jobs = -1,
                    silent = False,
                    verbosity = 0)        
                

print("starting training process..... \n") 
clf.fit(train_data, y_train)

starting training process..... 



/opt/conda/lib/python3.7/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, eval_metric='auc',
              gamma=0.05, gpu_id=-1, importance_type='gain',
              interaction_constraints='', learning_rate=0.02, max_delta_step=0,
              max_depth=8, min_child_weight=2, missing=nan,
              monotone_constraints='()', n_estimators=1350, n_jobs=-1,
              num_parallel_tree=1, random_state=0, reg_alpha=0, reg_lambda=1,
              scale_pos_weight=1, silent=False, subsample=0.8,
              tree_method='exact', validate_parameters=1, verbosity=0)

Sekarang kami menyimpan prediksi dalam file .csv.

In [67]:
sample_submission = pd.read_csv('../input/ieee-fraud-detection/sample_submission.csv', index_col='TransactionID')

sample_submission['isFraud'] = clf.predict_proba(test_data)[:,1]
sample_submission.to_csv('simple_xgboost.csv')

print("saving was successful!")

saving was successful!


In [68]:
# Read data simple_xgboost
simple_xgboost=pd.read_csv('simple_xgboost.csv')

In [69]:
# Menampilkan lima data pertama data simple_xgboot
simple_xgboost.head()

TransactionID   isFraud
0        3663549  0.005503
1        3663550  0.003363
2        3663551  0.007004
3        3663552  0.002290
4        3663553  0.001954

Modified by Elga Sania Rahman from https://www.kaggle.com/jonas0/ieee-fraud-detection-tutorial#Hello-and-welcome-to-my-IEEE-fraud-detection-tutorial!